include/picongpu/simulation/stage/SynchrotronRadiation.hpp

```diff
@@ -0,0 +1,252 @@
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with PIConGPU.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "picongpu/param/synchrotron.param"
+
+#include <pmacc/meta/ForEach.hpp>
+#include <pmacc/particles/traits/FilterByFlag.hpp>
+
+#include <cstdint>
+
+namespace picongpu::simulation::stage
+{
+    /** Functor for the stage of the PIC loop performing synchrotron radiation
+     *
+     * Only affects particle species with the Synchrotron attribute.
+     */
+    class SynchrotronRadiation
+    {
+    private:
+        /** exponential integration
+         *
+         * @brief
+         * 1) approximate function with exponential function by using the xLeft and xMiddle point to fit the exponent
+         * 2) integrate the exponential function on log scale,
+         *
+         * @note overestimates function below xMiddle but underestimates function above xMiddle leading to better fit
+         * overal (for bessel function).
+         *
+         * @param xLeft left point
+         * @param xMiddle middle point
+         * @param xRight right point
+         * @param yLeft function value at xLeft
+         * @param yMiddle function value at xMiddle
+         */
+        template<typename T_Number>
+        static T_Number integrateAsExponential(
+            T_Number const xLeft,
+            T_Number const xMiddle,
+            T_Number const xRight,
+            T_Number const yLeft,
+            T_Number const yMiddle)
+        {
+            if(xLeft == xMiddle)
+            {
+                throw "SynchrotronRadiation.hpp: integrateAsExponential(): xLeft and xMiddle cannot be the same.";
+            }
+            if(yLeft <= 0 || yMiddle <= 0)
+            {
+                throw "SynchrotronRadiation.hpp: integrateAsExponential(): yLeft and yMiddle must be positive.";
+            }
+            if(yLeft == yMiddle)
+            {
+                throw "SynchrotronRadiation.hpp: integrateAsExponential(): yLeft and yMiddle must not be equal.";
+            }
+
+            //! fitting function: y = a * e^(b * x)
+            T_Number const b = (math::log(yMiddle) - math::log(yLeft)) / (xMiddle - xLeft);
+            T_Number const a = yLeft / math::exp(b * xLeft);
+
+            T_Number const integral = (a / b) * (math::exp(b * xRight) - math::exp(b * xLeft));
+            return integral;
+        }
+
+        /** compute first synchrotorn function. See paper: "Extended particle-in-cell schemes for physics in
+         * ultrastrong laser fields: Review and developments" by A. Gonoskov et.Al.
+         *
+         * @param zq
+         *
+         * @returns zq * (integral of 2nd kind bessel function from zq to ~infinity)
+         */
+        static float_64 firstSynchrotronFunction(float_64 const zq)
+        {
+            //! for "FirstSynchrotronFunctionParams" struct
+            using namespace picongpu::particles::synchrotron::params;
```

Comment:
```suggestion
            using picongpu::particles::synchrotron::params::FirstSynchrotronFunctionParams;
```
if it's really only for this.

Reply:
done